azure/services/aso/service.go

```diff
@@ -83,6 +86,25 @@ func (s *Service[T, S]) Reconcile(ctx context.Context) error {
 		}
 	}
 
+	// If an error occurred during any create/update, don't delete any resources in case not all reconciled
```

Comment:
This comment is a bit weird here because it is describing the conditions in which we _don't_ actually do the following code block (or am I reading it wrong?)

Reply:
Correct. Would it make more sense to invert the language of the comment to say something like "only when no errors occur during any create/update, delete resources..."?

Reply:
Even that's sort of weird. My main confusion is: if we have an orphaned resource, why bother even reconciling at all? Wouldn't that be a preliminary short-circuit that would suggest that we should just garbage collect?

Reply:
This `Reconcile` function in particular doesn't reconcile an individual resource, but a group of them, e.g. all of the subnets defined for an AzureCluster.

> Wouldn't that be a preliminary short-circuit that would suggest that we should just garbage collect?

I think that's basically what this is already, where if no specs are defined, the loop immediately above has nothing to do and we skip straight here. Otherwise where should the short-circuit go?

Reply:
Right, I'm not suggesting it's incorrect, just perhaps a bit hard for a human to reason about.

Would this be better or worse?:

if no specs
  then cleanup, if listfunc is implemented, else exit early
else reconcile specs

Reply:
I think this is already doing the equivalent of that, just not in exactly that order (this reconciles first, then garbage collects).

Reply:
Agreed. If we reorder are we able to avoid weird (to a human) conditional composition like this?

`if (resultErr == nil || azure.IsOperationNotDoneError(resultErr)) && s.ListFunc != nil {`

Reply:
Yeah, it looks like mostly just swapping the order of the two blocks seems to eliminate the need for that. I'll make that change.